### PR TITLE
solution 21/07

### PR DESCRIPTION
**Pull request by VadimKadasiev**

### CODE REVIEW

File: src/main/java/book/store/controller/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping(value = "/categories")
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/controller/CategoryController.java]:
```suggestion
@RequestMapping("/categories")
```

File: src/main/java/book/store/dto/CreateBookRequestDto.java [hunk 2]
```diff
@@ -20,4 +21,6 @@ public class CreateBookRequestDto {
     private String description;
     @NotNull(message = "Cover_Image field can't be empty")
     private String coverImage;
+    @NotNull(message = "CategoryIds field can't be empty")
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/dto/CreateBookRequestDto.java]:
```suggestion
    @NotEmpty(message = "CategoryIds field can't be empty")
```

File: src/main/java/book/store/dto/CreateBookRequestDto.java [hunk 3]
```diff
@@ -20,4 +21,6 @@ public class CreateBookRequestDto {
     private String description;
     @NotNull(message = "Cover_Image field can't be empty")
     private String coverImage;
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/dto/CreateBookRequestDto.java]: Also, use `@NotBlank` for String

File: src/main/java/book/store/mapper/BookMapper.java [hunk 4]
```diff
@@ -2,13 +2,25 @@
...
+
     BookDto toDto(Book book);
 
     Book toModel(CreateBookRequestDto bookRequestDto);
+
+    BookDtoWithoutCategoryIds toDtoWithoutCategories(Book book);
+
+    @AfterMapping
+    default void setCategoryIds(@MappingTarget BookDto bookDto, Book book) {
+        bookDto.setCategoryIds(book.getCategories().stream().map(Category::getId).toList());
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/mapper/BookMapper.java]:
```suggestion
        bookDto.setCategoryIds(book.getCategories()
        .stream()
        .map(Category::getId)
        .toList());
```

File: src/main/java/book/store/model/Category.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+@Setter
+@Table(name = "categories")
+@SQLDelete(sql = "UPDATE categories SET is_deleted = true WHERE id = ?")
+@SQLRestriction(value = "is_deleted=false")
+public class Category {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private long id;
+
+    @Column(nullable = false, unique = true)
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/model/Category.java]:
```suggestion
    @Column(nullable = false)
```

File: src/main/java/book/store/repository/CategoryRepository.java [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+import book.store.model.Category;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface CategoryRepository extends JpaRepository<Category, Long> {
+    Category save(Category category);
+
+    Category getById(Long id);
+
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/repository/CategoryRepository.java]:
```suggestion
```

File: src/main/java/book/store/service/impl/CategoryServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+    @Override
+    public CategoryDto save(CategoryDto categoryDto) {
+        Category category = categoryMapper.toModel(categoryDto);
+        return categoryMapper.toDto(categoryRepository.save(category));
+    }
+
+    @Override
+    public List<CategoryDto> findAll(Pageable pageable) {
+        return categoryRepository.findAll(pageable).stream().map(categoryMapper::toDto).toList();
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/impl/CategoryServiceImpl.java]:
```suggestion
        return categoryRepository.findAll(pageable)
        .stream()
        .map(categoryMapper::toDto)
        .toList();
```

File: src/main/java/book/store/service/impl/CategoryServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    @Override
+    public CategoryDto getById(Long id) {
+        return categoryMapper.toDto(categoryRepository.findById(id).get());
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/impl/CategoryServiceImpl.java]: add .orElseThrow(()

File: src/main/java/book/store/service/impl/CategoryServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    @Override
+    public CategoryDto update(Long id, CategoryDto categoryDto) {
+        Category category = categoryMapper.toModel(categoryDto);
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/impl/CategoryServiceImpl.java]: First, find Category by id
Second, use MapStruct to update `mapper.updateCategoryFromDto(categoryDto, category);`

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Let's improve your implementation

File: src/main/java/book/store/service/impl/CategoryServiceImpl.java [hunk 10]
```diff
@@ -24,18 +24,23 @@ public CategoryDto save(CategoryDto categoryDto) {
...
+        return categoryMapper.toDto(categoryRepository.findById(id)
+                .orElseThrow(() -> new RuntimeException("Can't find category with ID " + id)));
     }
 
     @Override
     public CategoryDto update(Long id, CategoryDto categoryDto) {
-        Category category = categoryMapper.toModel(categoryDto);
-        category.setId(id);
+        Category category = categoryRepository.findById(id)
+                .orElseThrow(() -> new RuntimeException("Can't find category with ID " + id));
```
Comment (okuzan) [inline src/main/java/book/store/service/impl/CategoryServiceImpl.java]: please use EntityNotFoundException for such cases. Avoid overusing generic exceptions like runtime exception

**Review by okuzan — CHANGES_REQUESTED** [review]

File: src/main/java/book/store/model/Category.java [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+
+@Entity
+@Getter
+@Setter
+@Table(name = "categories")
+@SQLDelete(sql = "UPDATE categories SET is_deleted = true WHERE id = ?")
+@SQLRestriction(value = "is_deleted=false")
```
Comment (okuzan) [inline src/main/java/book/store/model/Category.java]:
```suggestion
@SQLDelete(sql = "UPDATE categories SET is_deleted = TRUE WHERE id = ?")
@SQLRestriction(value = "is_deleted = FALSE")
```

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 12]
```diff
@@ -10,4 +10,8 @@ databaseChangeLog:
   - include:
       file: db/changelog/changes/05-insert-roles.yaml
   - include:
-      file: db/changelog/changes/06-insert-admin.yaml
\ No newline at end of file
+      file: db/changelog/changes/06-insert-admin.yaml
+  - include:
+      file: db/changelog/changes/07-create-categories-table.yaml
+  - include:
+      file: db/changelog/changes/08-create-books-categories-table.yaml
```
Comment (okuzan) [inline src/main/resources/db/changelog/db.changelog-master.yaml]:
```suggestion
      file: db/changelog/changes/08-create-books-categories-table.yaml

```

**Review by okuzan — CHANGES_REQUESTED** [review]

File: src/main/java/book/store/controller/CategoryController.java [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/categories")
+public class CategoryController {
+    private final CategoryService categoryService;
+    private final BookService bookService;
+
+    @PreAuthorize("hasRole('ADMIN')")
+    @PostMapping
+    public CategoryDto createCategory(@RequestBody @Valid CategoryDto categoryDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/controller/CategoryController.java]: Why do you provide dto with Id in request to createCategory endpoint?

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comment

**Review by okuzan — APPROVED** [review]

**Pull request merged**